File: LeetCode/cpp/147.insertion-sort-list.cpp
/*
 * @lc app=leetcode id=147 lang=cpp
 *
 * [147] Insertion Sort List
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

// In-place insertion sort.
class Solution
{
public:
    ListNode *insertionSortList(ListNode *head)
    {
        if (head == nullptr)
            return nullptr;
        ListNode dummy(0);
        ListNode *cur = head;
        while (cur != nullptr)
        {
            ListNode *ele = cur;
            cur = cur->next;
            ListNode *prev = &dummy;
            while (prev->next != nullptr && prev->next->val <= ele->val)
            {
                prev = prev->next;
            }
            ele->next = prev->next;
            prev->next = ele;
        }
        return dummy.next;
    }
};


// Copy and sort. Just for comparing.
// Note that the problem is specified to insertion sort.
// ref:https://stackoverflow.com/questions/1525117/whats-the-fastst-algorithm-for-sorting-a-linked-list
class Solution_SortList
{
public:
    ListNode *insertionSortList(ListNode *head)
    {
        vector<int> nums;ListNode *cur = head;
        for (cur = head; cur != nullptr; cur = cur->next)
            nums.push_back(cur->val);

        sort(nums.begin(), nums.end());

        cur = head;
        for (int i = 0; i < nums.size(); ++i, cur = cur->next)
        {
            cur->val = nums[i];
        }
        return head;
    }
}

File: LeetCode/java/147.insertion-sort-list.java
/*
 * @lc app=leetcode id=147 lang=java
 *
 * [147] Insertion Sort List
 *
 * https://leetcode.com/problems/insertion-sort-list/description/
 *
 * algorithms
 * Medium (40.77%)
 * Likes:    933
 * Dislikes: 663
 * Total Accepted:    222.2K
 * Total Submissions: 503.3K
 * Testcase Example:  '[4,2,1,3]'
 *
 * Sort a linked list using insertion sort.
 * 
 * 
 * 
 * 
 * 
 * A graphical example of insertion sort. The partial sorted list (black)
 * initially contains only the first element in the list.
 * With each iteration one element (red) is removed from the input data and
 * inserted in-place into the sorted list
 * 
 * 
 * 
 * 
 * 
 * Algorithm of Insertion Sort:
 * 
 * 
 * Insertion sort iterates, consuming one input element each repetition, and
 * growing a sorted output list.
 * At each iteration, insertion sort removes one element from the input data,
 * finds the location it belongs within the sorted list, and inserts it
 * there.
 * It repeats until no input elements remain.
 * 
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: 4->2->1->3
 * Output: 1->2->3->4
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: -1->5->3->4->0
 * Output: -1->0->3->4->5
 * 
 * 
 */

// @lc code=start
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode() {} ListNode(int val) { this.val = val; } ListNode(int val,
 * ListNode next) { this.val = val; this.next = next; } }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        if (head == null || head.next == null)
            return head;
        ListNode dummy = new ListNode();

        while (head != null) {
            ListNode ele = head;

            head = head.next;

            // Find the insertion point and insert it.
            ListNode cur = dummy;
            while (cur.next != null && cur.next.val <= ele.val)
                cur = cur.next;

            // Insert the element
            ele.next = cur.next;
            cur.next = ele;

        }
        return dummy.next;
    }
}
// @lc code=end
